File: packages/sdk/src/harness/core/circuit_breaker.py
"""
Circuit breaker for detecting and preventing infinite loops.

Simple, scalable approach following Bitter Lesson principles:
- Simple rules over complex heuristics
- Trust the model to self-correct
- Only intervene on obvious problems (same tool + same args repeated)
"""

from __future__ import annotations

from collections import Counter
from dataclasses import dataclass, field
from datetime import datetime, timedelta
from typing import Any
from enum import Enum


class CircuitState(Enum):
    """Circuit breaker states."""
    CLOSED = "closed"      # Normal operation
    OPEN = "open"          # Blocking calls
    HALF_OPEN = "half_open"  # Testing if recovered


@dataclass
class CircuitBreakerConfig:
    """Configuration for circuit breaker.

    Following Bitter Lesson: keep it simple, trust the model.
    """
    # Same tool + args threshold
    # Only trigger when calling same tool with same arguments repeatedly
    # This is the simplest, most reliable detection
    same_args_threshold: int = 3

    # Error threshold
    error_threshold: int = 5
    error_window_seconds: int = 60

    # Recovery
    recovery_timeout_seconds: int = 30
    half_open_max_calls: int = 1


class CircuitBreaker:
    """
    Circuit breaker to detect and prevent infinite loops.

    Simple approach:
    - Detect when same tool is called with same arguments repeatedly
    - Detect when too many errors occur in a short time

    We don't try to be "smart" about detecting complex patterns.
    Trust the model (via system prompt) to know when to stop.
    """

    def __init__(self, config: CircuitBreakerConfig | None = None):
        self.config = config or CircuitBreakerConfig()
        self._state = CircuitState.CLOSED

        # Track tool:args combination (simple and effective)
        self._tool_args_counter: Counter[str] = Counter()

        # Track errors
        self._error_times: list[datetime] = []

        # Track half-open state
        self._open_time: datetime | None = None
        self._half_open_calls = 0

        # Track why circuit opened
        self._open_reason: str | None = None

    def record_call(self, tool_name: str, arguments: dict[str, Any]) -> None:
        """
        Record a tool call for pattern detection.

        Simple: just count tool:args combinations.
        """
        args_key = self._make_args_key(tool_name, arguments)
        self._tool_args_counter[args_key] += 1

        # Check if we should open circuit
        self._check_patterns()

    def _make_args_key(self, tool_name: str, arguments: dict[str, Any]) -> str:
        """Create a hashable key for tool + args combination."""
        args_hash = self._hashable_args(arguments)
        return f"{tool_name}:{args_hash}"

    def record_error(self, error: Exception | None = None) -> None:
        """Record an error during tool execution."""
        self._error_times.append(datetime.now())

        # Clean old errors outside window
        cutoff = datetime.now() - timedelta(seconds=self.config.error_window_seconds)
        self._error_times = [t for t in self._error_times if t > cutoff]

        # Check if error threshold reached
        if len(self._error_times) >= self.config.error_threshold:
            self._open_reason = (
                f"{len(self._error_times)} errors in last "
                f"{self.config.error_window_seconds} seconds"
            )
            self._open()

    def record_success(self) -> None:
        """Record a successful tool execution."""
        if self._state == CircuitState.HALF_OPEN:
            self._close()

    def is_open(self) -> bool:
        """Check if circuit is open (should block calls)."""
        if self._state == CircuitState.CLOSED:
            return False

        if self._state == CircuitState.OPEN:
            if self._open_time:
                elapsed = (datetime.now() - self._open_time).total_seconds()
                if elapsed >= self.config.recovery_timeout_seconds:
                    self._half_open()
                    return False
            return True

        if self._half_open_calls >= self.config.half_open_max_calls:
            return True

        return False

    def get_reason(self) -> str | None:
        """Get the reason why circuit is open."""
        if self._state == CircuitState.CLOSED:
            return None
        return self._open_reason or "Circuit breaker is open"

    def reset(self) -> None:
        """Reset the circuit breaker to closed state."""
        self._close()
        self._tool_args_counter.clear()
        self._error_times.clear()
        self._open_reason = None

    def _check_patterns(self) -> None:
        """Check for problematic patterns - keep it simple."""
        # Only check: same tool + same args repeated
        for args_key, count in self._tool_args_counter.items():
            if count >= self.config.same_args_threshold:
                tool_name = args_key.split(":", 1)[0]
                self._open_reason = (
                    f"Tool '{tool_name}' with same arguments called {count} times "
                    f"(threshold: {self.config.same_args_threshold})"
                )
                self._open()
                return

    def _hashable_args(self, args: dict[str, Any]) -> tuple:
        """Convert args dict to hashable tuple for comparison."""
        items = []
        for k, v in sorted(args.items()):
            if isinstance(v, dict):
                v = self._hashable_args(v)
            elif isinstance(v, list):
                v = tuple(v)
            items.append((k, v))
        return tuple(items)

    def _open(self) -> None:
        """Open the circuit."""
        self._state = CircuitState.OPEN
        self._open_time = datetime.now()
        self._half_open_calls = 0

    def _half_open(self) -> None:
        """Move to half-open state."""
        self._state = CircuitState.HALF_OPEN
        self._half_open_calls = 0

    def _close(self) -> None:
        """Close the circuit."""
        self._state = CircuitState.CLOSED
        self._open_time = None
        self._half_open_calls = 0

    @property
    def state(self) -> CircuitState:
        """Get current circuit state."""
        return self._state

    @property
    def stats(self) -> dict[str, Any]:
        """Get circuit breaker statistics."""
        return {
            "state": self._state.value,
            "recent_errors": len(self._error_times),
            "tool_args_counter": dict(self._tool_args_counter.most_common(5)),
        }

File: packages/sdk-java/harness-sdk-core/src/main/java/com/harness/core/CircuitBreaker.java
package com.harness.core;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Circuit breaker to detect and prevent infinite loops.
 *
 * Simple approach:
 * - Detect when same tool is called with same arguments repeatedly
 * - Detect when too many errors occur in a short time
 *
 * We don't try to be "smart" about detecting complex patterns.
 * Trust the model (via system prompt) to know when to stop.
 *
 * Example:
 * <pre>
 * CircuitBreaker cb = new CircuitBreaker();
 *
 * // Before tool execution
 * if (cb.isOpen()) {
 *     throw new CircuitBreakerException(cb.getReason());
 * }
 *
 * // Record call
 * cb.recordCall("read", Map.of("path", "/tmp/file.txt"));
 *
 * // On error
 * cb.recordError(e);
 *
 * // On success
 * cb.recordSuccess();
 * </pre>
 */
public class CircuitBreaker {

    private static final Logger logger = LoggerFactory.getLogger(CircuitBreaker.class);

    private final CircuitBreakerConfig config;
    private CircuitState state = CircuitState.CLOSED;

    // Track tool:args combination (simple and effective)
    private final Map<String, Integer> toolArgsCounter = new HashMap<>();

    // Track errors
    private final List<Instant> errorTimes = new ArrayList<>();

    // Track half-open state
    private Instant openTime = null;
    private int halfOpenCalls = 0;

    // Track why circuit opened
    private String openReason = null;

    public CircuitBreaker() {
        this(CircuitBreakerConfig.defaults());
    }

    public CircuitBreaker(CircuitBreakerConfig config) {
        this.config = config;
    }

    /**
     * Record a tool call for pattern detection.
     *
     * Simple: just count tool:args combinations.
     */
    public void recordCall(String toolName, Map<String, Object> arguments) {
        String argsKey = makeArgsKey(toolName, arguments);
        toolArgsCounter.merge(argsKey, 1, Integer::sum);

        // Check if we should open circuit
        checkPatterns();
    }

    /**
     * Record an error during tool execution.
     */
    public void recordError(Exception error) {
        errorTimes.add(Instant.now());

        // Clean old errors outside window
        Instant cutoff = Instant.now().minus(config.errorWindowSeconds(), ChronoUnit.SECONDS);
        errorTimes.removeIf(t -> t.isBefore(cutoff));

        // Check if error threshold reached
        if (errorTimes.size() >= config.errorThreshold()) {
            openReason = String.format("%d errors in last %d seconds",
                errorTimes.size(), config.errorWindowSeconds());
            open();
        }
    }

    /**
     * Record a successful tool execution.
     */
    public void recordSuccess() {
        if (state == CircuitState.HALF_OPEN) {
            close();
        }
    }

    /**
     * Check if circuit is open (should block calls).
     */
    public boolean isOpen() {
        if (state == CircuitState.CLOSED) {
            return false;
        }

        if (state == CircuitState.OPEN) {
            if (openTime != null) {
                long elapsed = ChronoUnit.SECONDS.between(openTime, Instant.now());
                if (elapsed >= config.recoveryTimeoutSeconds()) {
                    halfOpen();
                    return false;
                }
            }
            return true;
        }

        // HALF_OPEN state
        if (halfOpenCalls >= config.halfOpenMaxCalls()) {
            return true;
        }

        return false;
    }

    /**
     * Get the reason why circuit is open.
     */
    public String getReason() {
        if (state == CircuitState.CLOSED) {
            return null;
        }
        return openReason != null ? openReason : "Circuit breaker is open";
    }

    /**
     * Reset the circuit breaker to closed state.
     */
    public void reset() {
        close();
        toolArgsCounter.clear();
        errorTimes.clear();
        openReason = null;
    }

    /**
     * Get current circuit state.
     */
    public CircuitState getState() {
        return state;
    }

    /**
     * Get circuit breaker statistics.
     */
    public Map<String, Object> getStats() {
        Map<String, Object> stats = new LinkedHashMap<>();
        stats.put("state", state.name());
        stats.put("recentErrors", errorTimes.size());

        // Get top 5 most common tool:args combinations
        Map<String, Integer> topCalls = new LinkedHashMap<>();
        toolArgsCounter.entrySet().stream()
            .sorted(Map.Entry.<String, Integer>comparingByValue().reversed())
            .limit(5)
            .forEach(e -> topCalls.put(e.getKey(), e.getValue()));
        stats.put("toolArgsCounter", topCalls);

        return stats;
    }

    // === Private Methods ===

    private void checkPatterns() {
        // Only check: same tool + same args repeated
        for (Map.Entry<String, Integer> entry : toolArgsCounter.entrySet()) {
            if (entry.getValue() >= config.sameArgsThreshold()) {
                String toolName = entry.getKey().split(":", 2)[0];
                openReason = String.format(
                    "Tool '%s' with same arguments called %d times (threshold: %d)",
                    toolName, entry.getValue(), config.sameArgsThreshold()
                );
                logger.warn("Circuit breaker opening: {}", openReason);
                open();
                return;
            }
        }
    }

    private String makeArgsKey(String toolName, Map<String, Object> arguments) {
        String argsHash = hashableArgs(arguments);
        return toolName + ":" + argsHash;
    }

    private String hashableArgs(Map<String, Object> args) {
        if (args == null || args.isEmpty()) {
            return "{}";
        }

        // Simple string representation for hashing
        StringBuilder sb = new StringBuilder();
        sb.append("{");
        boolean first = true;
        for (Map.Entry<String, Object> entry : args.entrySet()) {
            if (!first) sb.append(",");
            first = false;
            sb.append(entry.getKey());
            sb.append("=");
            Object value = entry.getValue();
            if (value instanceof Map) {
                sb.append(hashableArgs((Map<String, Object>) value));
            } else if (value instanceof List) {
                sb.append(value.toString());
            } else {
                sb.append(value != null ? value.toString() : "null");
            }
        }
        sb.append("}");
        return sb.toString();
    }

    private void open() {
        state = CircuitState.OPEN;
        openTime = Instant.now();
        halfOpenCalls = 0;
    }

    private void halfOpen() {
        state = CircuitState.HALF_OPEN;
        halfOpenCalls = 0;
        logger.info("Circuit breaker entering half-open state");
    }

    private void close() {
        state = CircuitState.CLOSED;
        openTime = null;
        halfOpenCalls = 0;
        logger.info("Circuit breaker closed");
    }
}
